### PR TITLE
Copter: Return currently assigned local target

### PR DESCRIPTION
**Pull request by rajat2004**
When an operator sends the SET_POSITION_TARGET_LOCAL_NED message there exists no acknowledgement of the command. This commit adds functionality for the copter to return the assigned local target. Please check Issue #8686 for more details

### CODE REVIEW

**Comment by rmackay9** [comment]
Hi, thanks for this contribution.  I'm not sure this is the best way though I'm afraid.  This is a one-time send which could also very easily be lost (MAVLink doesn't guarantee delivery).  I think the way it should be done is if the vehicle is using the target-local-ned then it should send this position-target message regularly.  I'm not sure how that can be done though.

As a side note the commit should have "Copter:" prefixed to the beginning.

**Comment by WickedShell** [comment]
@rmackay9 I'd disagree with that actually. We probably should send it continuously as well, but there are a lot of precedents to send it as a response to an external command with MAVLink. If you don't ack the message in some way for the GCS then the GCS/external systems life gets a lot harder. The periodic resend is needed for transport reasons as you said, but we should still try and respond as quickly as possible.

**Comment by rmackay9** [comment]
@WickedShell, but we both agree that the one-time send is not sufficient right?

**Comment by WickedShell** [comment]
@rmackay9 yeah, but it's at least a start and a needed part of the solution :)

**Comment by khancyr** [comment]
maybe we should restrict that to position only. On velocity control at > 30Hz, I don't see the point to get back what was send.
even with position it will be more useful to send it regulary and not only once

**Comment by peterbarker** [comment]
If we do send a response straight away, it should not simply be echoing back what we received in the packet; we may have ignored several fields, had to limit the range of a field, that sort of thing.

The spec says we should echo back, however.

So... if the packet says, "move your Rover down 1m in NED" - should you echo that position target back to the GCS verbatim or clamp the vertical movement to terrain altitude at destination?  Do we keep around "true position target" and "position target we're telling the GCS"?!

**Comment by rajat2004** [comment]
Thanks for the responses.
So to summarize, 
1. There should be a periodic resend of the position_target message rather than one-time response.
2. This could be done for position only, not required when controlling the drone with velocity control at high rate
3. Should the response echo the message recieved verbatim or not.

For the 1st point, I'm looking into using SET_MESSAGE_INTERVAL for sending the message continously.

As for the 3rd, I'm not sure if there's any way other than to send the exact message recieved.  What could be done is explicitly setting the return values according to the `ignore` fields but I think that would have already been taken care of by the operator.

If there's any better way, please let me know.

@rmackay9 I thought that was only for PR topics. I'll definitely keep that in mind form now on.

**Comment by rajat2004** [comment]
I looked at the following methods for sending the message regularly
1. Using SET_MESSSAGE_INTERVAL
2. SRx_Params
3. Scheduling the message

I think that the biggest problem is that the target data isn't stored anywhere and so, there can't be something like a function to send the message, otherwise, it could have been used in the scheduler. With the SRx params, the data streams are defined and don't support this message. Even with using  set_message_interval, only the msg_id is encoded in the commands, not the message itself, so it won't be able to get the data. 
Are any of the above inferences wrong, and are there any other ways to regularly send a message?

**Comment by peterbarker** [comment]
On Sat, 9 Mar 2019, Rajat Singhal wrote:

> I looked at the following methods for sending the message regularly
>
>  1. Using SET_MESSSAGE_INTERVAL
>  2. SRx_Params
>  3. Scheduling the message
> 
> I think that the biggest problem is that the target data isn't stored anywhere and so, there can't be something like a function to send the message,
> otherwise, it could have been used in the scheduler. With the SRx params, the data streams are defined and don't support this message. Even with using
> set_message_interval, only the msg_id is encoded in the commands, not the message itself, so it won't be able to get the data.
> Are any of the above inferences wrong, and are there any other ways to regularly send a message?

You need to delve a little deeper into the way send_message works :-)

The ID is used just for queueing.  There's a switch in try_send_message on 
that ID which tries to send a mavlink message based on the ID passed in.

You can depend on SET_MESSAGE_INTERVAL without relying putting things into 
the SRx_ parameters.

Gathering the data to send the message is most of the fun with this PR :-)

It may be similar to the way we do SEND_POSITION_TARGET_GLOBAL_INT and 
NAV_CONTROLLER_OUTPUT

**Comment by rajat2004** [comment]
Thanks a lot for the tip. You were right, following the functions across files and libraries is fun.

By following the functions which implement the SET_POSITION_TARGET and the SEND_POSITION_TARGET_GLOBAL_INT, I was able to find where the current target data is stored,  in the [mode_guided.cpp](https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/mode_guided.cpp).

So first I'll create a function to send the POSITION_TARGET_LOCAL_NED and then go ahead from there. But the frame type will have to be a default one similar to how it's done in [send_position_target_global_int()](https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/GCS_Mavlink.cpp#L106)

And then I'll add it in [GCS_Common.cpp similar to position_target_global_int](https://github.com/ArduPilot/ardupilot/blob/d48eab893a3dd38929d9271de59e0145f1ae3a19/libraries/GCS_MAVLink/GCS_Common.cpp#L4060).

**Comment by rajat2004** [comment]
Really sorry for being out of contact for so long

I have completely removed the previous implementation and made it into a separate function. Added the message in the appropriate places for using SET_MESSAGE_INTERVAL for POSITION_TARGET_LOCAL_NED.
So the operator can use SET_MESSAGE_INTERVAL to check the current target.

Currently, LOCAL_NED is exactly similar to GLOBAL_INT and behaves in the same manner. 

Next step:- Returning the assigned target at regular intervals only when controlling using position 

Please have a look and suggest if any more changes are needed.

File: ArduCopter/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -118,6 +118,29 @@ void GCS_MAVLINK_Copter::send_position_target_global_int()
         0.0f); // yaw_rate
 }
 
+void GCS_MAVLINK_Copter::send_position_target_local_ned()
+{
+    const Vector3f target_pos = copter.wp_nav->get_wp_destination();
```
Comment (peterbarker) [inline ArduCopter/GCS_Mavlink.cpp]: What are these going to return when we're in `STABILIZE` mode?  What would someone reviewing a log with such a section in it think?

Comment (rajat2004) [reply ArduCopter/GCS_Mavlink.cpp]: Added check for whether copter is in Guided mode or not

**Comment by rajat2004** [comment]
I have listed the cases in which to send the POSITION_TARGET messages

1. Position control - Immediately respond with the target as well as set the message to be sent at regular intervals using SET_MESSAGE_INTERVAL
2. Velocity control -  Send the message at regular intervals also, but limit the frequency of the message.

Is there anything else which should be added or changed? What would be a proper frequency for the messages?

**Comment by rajat2004** [comment]
Anyone?
This can be added for POSITION_TARGET_GLOBAL_INT also

**Comment by rajat2004** [comment]
I have listed the cases in which to send the POSITION_TARGET messages

   1.  Position control - Immediately respond with the target as well as set the message to be sent at regular intervals using SET_MESSAGE_INTERVAL
   2. Velocity control - Send the message at regular intervals also, but limit the frequency of the message.

Is there anything else which should be added or changed? What would be a proper frequency for the messages?

**Comment by rajat2004** [comment]
I have listed the cases in which to send the POSITION_TARGET messages

   1.  Position control - Immediately respond with the target as well as set the message to be sent at regular intervals using SET_MESSAGE_INTERVAL
   2. Velocity control - Send the message at regular intervals also, but limit the frequency of the message.

Is there anything else which should be added or changed? What would be a proper frequency for the messages?
This can be added for POSITION_TARGET_GLOBAL_INT also

**Comment by rajat2004** [comment]
Can someone have a look at this and suggest if this is the proper way or not?
Also, should this additional functionality of sending the POSITION_TARGET messages be added in this same PR or a separate one?

**Comment by rajat2004** [comment]
Can someone have a look at this and suggest if this is the proper way or not?
Also, should this additional functionality of sending the POSITION_TARGET messages be added in this same PR or a separate one?

**Comment by rajat2004** [comment]
I had written my function in a similar way to the ```send_position_target_global_int()``` function in which only the target location is being sent with a fixed typemask.

Added code for sending the appropriate typemask, velocity and position target.
Suggestions for cleanup and comments are welcome.

Regarding the commit message, I think you are referring to the first one? I now know that it wasn't proper, but can I change it after pushing?

**Comment by rajat2004** [comment]
A bit of possible cleanup from my side 
Initialize ```target_pos``` as
```
Vector3f target_pos = Vector3f(0.0f, 0.0f, 0.0f);
```

Another thing which can be done is to convert into centimeters when we get the data itself in the appropriate condition
```
target_pos = copter.wp_nav->get_wp_destination() * 0.01f;
```

I'm not sure if this is proper according to Ardupilot's Coding style

File: ArduCopter/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -118,6 +118,49 @@ void GCS_MAVLINK_Copter::send_position_target_global_int()
...
+    
+    const GuidedMode guided_mode = copter.mode_guided.mode();
+    Vector3f target_pos;
+    Vector3f target_vel;
+    uint16_t type_mask;
+
+    if (guided_mode == Guided_WP) {
+        type_mask = 0x0FF8;
+        target_pos = copter.wp_nav->get_wp_destination();
+        target_vel = Vector3f(0.0f, 0.0f, 0.0f);
```
Comment (peterbarker) [inline ArduCopter/GCS_Mavlink.cpp]: These are implicitly zero; you don't need to do this call at all.

**Comment by peterbarker** [comment]
On Mon, 25 Mar 2019, Rajat Singhal wrote:

> Regarding the commit message, I think you are referring to the first one?
> I now know that it wasn't proper, but can I change it after pushing?

Yes!  You can force-push up your branch as much as you like.

File: ArduCopter/GCS_Mavlink.cpp (hunk 2 above)
Comment (rajat2004) [reply ArduCopter/GCS_Mavlink.cpp]: Okay, thanks for the info. I'll clean this up. Anything else which I should do?

**Comment by rajat2004** [comment]
> Yes! You can force-push up your branch as much as you like.

I'll do this once the PR is good enough for merge

File: ArduCopter/GCS_Mavlink.cpp (hunk 2 above)
Comment (rajat2004) [reply ArduCopter/GCS_Mavlink.cpp]: Cleaned up the function, added comments also.
 
> Another thing which can be done is to convert into centimeters when we get the data itself in the appropriate condition
> 
> ```
> target_pos = copter.wp_nav->get_wp_destination() * 0.01f;
> ```

Should I add this also?

**Comment by rajat2004** [comment]
Should I squash any commits?

**Comment by rajat2004** [comment]
Any idea as to what could be done to fix the Travis build?
It's failing a ArduSub test which I haven't touched

File: ArduCopter/GCS_Mavlink.cpp (hunk 2 above)
Comment (peterbarker) [reply ArduCopter/GCS_Mavlink.cpp]: Yes, definitely.  No unnecessary work, and sometime we might be clever about optimising that multiply.

**Comment by peterbarker** [comment]
Try rebasing this on master.

You should only have a single commit involved in this, so squash everything together, please.

File: ArduCopter/GCS_Mavlink.cpp (hunk 2 above)
Comment (rajat2004) [reply ArduCopter/GCS_Mavlink.cpp]: Done

**Comment by rajat2004** [comment]
Rebased the branch and squashed all the commits.

**Comment by rajat2004** [comment]
> I have listed the cases in which to send the POSITION_TARGET messages
> 
>     1. Position control - Immediately respond with the target as well as set the message to be sent at regular intervals using SET_MESSAGE_INTERVAL
> 
>     2. Velocity control - Send the message at regular intervals also, but limit the frequency of the message.
> 
> 
> The above will, of course, be carried out only if the operator hasn't requested the POSITION_TARGET messages himself using SET_MESSAGE_INTERVAL or hasn't disabled it
> 
> Is there anything else which should be added or changed? What would be a proper frequency for the messages?
> This can be added for POSITION_TARGET_GLOBAL_INT also

Is all this required, or just keep it as it is?
The operator can decide whether he needs the info or not by using SET_MESSAGE_INTERVAL

**Comment by rajat2004** [comment]
This function which I have added for sending POSITION_TARGET_LOCAL_NED is missing from all the platforms, is something like this required for them also?
Maybe not in Plane since it only uses the altitude and sends a response to GCS also
In APMRover and ArduSub, there's also no way to send the global target also.

**Comment by peterbarker** [comment]
On Wed, 27 Mar 2019, Rajat Singhal wrote:

> This function which I have added for sending POSITION_TARGET_LOCAL_NED is missing from all the platforms, is something like this required for them also?
> Maybe not in Plane since it only uses the altitude and sends a response to GCS also
> In APMRover and ArduSub, there's also no way to send the global target also.

Let's focus on getting it into Copter first :-)

If you *did* want to experiment further - why not find a convenient GCS 
(e.g. MAVProxy - easy to hack on) and actually *use* the message in the 
display in some way.

If you're feeling *really* adventurous, drawing desired/actual velocity 
vectors in the 3D Morse SITL backend (when in a velocity rather than 
waypoint regime) would be cool :-)

**Comment by rajat2004** [comment]
What all further improvements need to be done for merging this into Copter?

The part about Mavproxy sounds cool.I'll definitely experiment with that

**Comment by rajat2004** [comment]
> > I have listed the cases in which to send the POSITION_TARGET messages
> > ```
> > 1. Position control - Immediately respond with the target as well as set the message to be sent at regular intervals using SET_MESSAGE_INTERVAL
> > 
> > 2. Velocity control - Send the message at regular intervals also, but limit the frequency of the message.
> > ```
> > The above will, of course, be carried out only if the operator hasn't requested the POSITION_TARGET messages himself using SET_MESSAGE_INTERVAL or hasn't disabled it
> > Is there anything else which should be added or changed? What would be a proper frequency for the messages?
> > This can be added for POSITION_TARGET_GLOBAL_INT also
> 
> Is all this required, or just keep it as it is?
> The operator can decide whether he needs the info or not by using SET_MESSAGE_INTERVAL

@peterbarker ?

**Comment by peterbarker** [comment]
@rajat2004 I've asked @rmackay9 to have another look at this for you.  He's the one that hsa to be happy for this to be merged :-)

One thing that *I* would really like to see what be something in our autotest suite to ensure we're getting the correct messages back.  There are several Copter tests which send through set-position-target-local-ned and would reasonably expect a reply - 'though our test suite isn't exhaustive in terms of the sorts of messages we might send.

https://github.com/ardupilot/ardupilot/blob/master/Tools/autotest/arducopter.py#L2067

**Comment by rmackay9** [comment]
So to be clear the POSITION_TARGET_LOCAL_NED is only sent if someone has specifically requested them using the SET_MESSAGE_INTERVAL or REQUEST_MESSAGE message right?

These messages are not included in any of the current streams right?

If "yes" then it looks good to me.  thanks!

**Comment by peterbarker** [comment]
@rajat2004 Gah!  I've given you a bad direction here.  There should be two commits - one for Copter, one for the GCS directory.  If you like I can split them again for you and push over your commit here.  Let me know.

BTW, new is if you write the autotests this will get merged - otherwise more discussion/testing :-)

If you can run sim_vehicle.py you can run autotest.  An example would be:

`./Tools/autotest/autotest.py build.ArduCopter fly.ArduCopter.GuidedSubModeChange`

Let me know if you're unwilling/unable to write the tests and I'll do something...

**Comment by rajat2004** [comment]
> So to be clear the POSITION_TARGET_LOCAL_NED is only sent if someone has specifically requested them using the SET_MESSAGE_INTERVAL or REQUEST_MESSAGE message right?
> 
> These messages are not included in any of the current streams right?

Yes, that's the current state.



> There should be two commits - one for Copter, one for the GCS directory. If you like I can split them again for you and push over your commit here. Let me know.
> 

So I should split this into 2 commits and then force-push them?
I'll try it, if it doesn't work out then I'll inform

> BTW, new is if you write the autotests this will get merged - otherwise more discussion/testing :-)
> 
> If you can run sim_vehicle.py you can run autotest. An example would be:
> 
> `./Tools/autotest/autotest.py build.ArduCopter fly.ArduCopter.GuidedSubModeChange`
> 
> Let me know if you're unwilling/unable to write the tests and I'll do something...

I didn't understand this. So I need to write some tests for checking whether the right position target messages are being recieved?

**Comment by rajat2004** [comment]
> So to be clear the POSITION_TARGET_LOCAL_NED is only sent if someone has specifically requested them using the SET_MESSAGE_INTERVAL or REQUEST_MESSAGE message right?
> 
> These messages are not included in any of the current streams right?

Yes, that's the current state.



> There should be two commits - one for Copter, one for the GCS directory. If you like I can split them again for you and push over your commit here. Let me know.
> 

So I should split this into 2 commits and then force-push them?
I'll try it, if it doesn't work out then I'll inform

> BTW, new is if you write the autotests this will get merged - otherwise more discussion/testing :-)
> 
> If you can run sim_vehicle.py you can run autotest. An example would be:
> 
> `./Tools/autotest/autotest.py build.ArduCopter fly.ArduCopter.GuidedSubModeChange`
> 
> Let me know if you're unwilling/unable to write the tests and I'll do something...

I didn't understand this. So I need to write some tests for checking whether the right position target messages are being recieved? But won't that require this PR to be merged first just to receive the messages

@peterbarker @rmackay9 ??

**Comment by rajat2004** [comment]
Commit spit into 2, one for GCS_Mavlink lib and for Copter
Is this fine @peterbarker?

**Comment by peterbarker** [comment]
@rajat2004 Merged this, thanks!

Now, on the testing...

This PR didn't need to be merged for tests to be written.  It would have just been an additional patch on top.  But your description of the tests apart from that is accurate - I highlighted a spot in the code where we could add a check that when you've asked for them, and it is appropriate for the autopilot to send them, we get these messages emitted.  Are you interested in writing such a test?

**Comment by rajat2004** [comment]
@peterbarker Thanks a lot for merging this!! Feels great to have contributed something
 
Regarding the test, I'm definitely interested. I'll have a look around, see if I'll be able to.

Are there any already written tests for checking if the correct message is being received after using SET_MESSAGE_INTERVAL? I had a look through the tests but I only found the following test for checking the message rate, not the content of the message itself
https://github.com/ArduPilot/ardupilot/blob/3c432cf25caeda152767473c5b4c498d8bfc887a/Tools/autotest/common.py#L2255

**Comment by rajat2004** [comment]
@peterbarker Thanks a lot for merging this!! Feels great to have contributed something
 
Regarding the test, I'm definitely interested. I'll have a look around, see if I'll be able to.

Are there any already written tests for checking if the correct message is being received after using SET_MESSAGE_INTERVAL? I had a quick look through the tests but I only found the following test for checking the message rate, not the content of the message itself
https://github.com/ArduPilot/ardupilot/blob/3c432cf25caeda152767473c5b4c498d8bfc887a/Tools/autotest/common.py#L2255

**Comment by peterbarker** [comment]
So you'll need to do a set-message-interval in one of the tests that does a set-position-target-local-ned.

Immediately after sending the message you should be able to do a blocking recv_match to get an instance of your return message, which you can then interrogate to make sure it looks good.

This might warrant an entirely new test - they're not hard to write :-)

**Comment by rajat2004** [comment]
Okay, so I do a Set-Message-Interval for set-position-local-target-ned and use recv_match to get a single message.
Then check the rate, type_mask, target position values, etc.
This could be done for all 3 modes in Guided - Pos, Vel & PosVel.

I think a new test would be better, won't interfere with the others?
Is this fine @peterbarker 

Also, will this function  ```send_posiiton_target_local_ned()``` which was added in Copter be useful for other platforms? Maybe not for Plane since it only uses the altitude and sends a response also but for the others?

**Comment by peterbarker** [comment]
On Fri, 29 Mar 2019, Rajat Singhal wrote:

> 
> Okay, so I do a Set-Message-Interval for set-position-local-target-ned and use recv_match to get a single message.
> Then check the rate, type_mask, target position values, etc.
> This could be done for all 3 modes in Guided - Pos, Vel & PosVel.
> 
> I think a new test would be better, won't interfere with the others?
> Is this fine @peterbarker

Sounds good, but you probably don't need to check the rate - just that the 
messages do what they're supposed to :-)

You could also add a test that we don't get the messages outside guided 
mode, even when we send in set-position-target-local-ned.

> Also, will this function send_position_target_local_ned() which was 
added in Copter be useful for other platforms? Maybe not for Plane since it only uses
> the altitude and sends a response also but for the others?

Test first :-)

But yes, definitely for the other Rover.

There are a slew of these messages that would be nice to get out - 
ATTITUDE_TARGET, for example.  Because we can now have messages that we 
can emit but don't send out all the time they're relatively cheap to add - 
but we really should motivate this by showing a GCS (or other client) 
actually using the data (as I've previously mentioned).  An example of 
where we've done this is NAV_CONTROLLER_OUTPUT, where the GCS often draws 
a vector from our current position to the target position.

**Comment by rajat2004** [comment]
I'll look into Mavproxy maybe displaying the velocity vector since it already displays the vector from the current position to the target position for local targets, after writing the tests and maybe even adding this for Rover

**Comment by rajat2004** [comment]
@peterbarker @rmackay9 I have opened #10955 for adding the tests as discussed above. Please a look at it.

**Pull request merged**